src/wp-includes/block-template-utils.php

```diff
@@ -549,15 +549,18 @@ function _build_block_template_result_from_file( $template_file, $template_type
 		$template->area = $template_file['area'];
 	}
 
-	$before_block_visitor = '_inject_theme_attribute_in_template_part_block';
+	$before_block_visitor = ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ? '_inject_theme_attribute_in_template_part_block' : null;
 	$after_block_visitor  = null;
 	$hooked_blocks        = get_hooked_blocks();
 	if ( ! empty( $hooked_blocks ) || has_filter( 'hooked_block_types' ) ) {
 		$before_block_visitor = make_before_block_visitor( $hooked_blocks, $template );
 		$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $template );
 	}
-	$blocks            = parse_blocks( $template_content );
-	$template->content = traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
+	if ( ! is_null( $before_block_visitor ) || ! is_null( $after_block_visitor ) ) {
```

Comment:
Nit-pick: We could use the simpler `null` equality checks here.
```suggestion
	if ( null !== $before_block_visitor || null !== $after_block_visitor ) {
```

Reply:
Simpler means faster? I coded primarily JavaScript for too long and I’m relearning PHP 😅

Reply:
I guess comparing variables is simpler than calling a function, that's all I'm saying.

But since you asked, I got curious: https://3v4l.org/mj5rq 😂 

Seems there is an _extremely_ tiny performance benefit to it too. Really irrelevant, so not a blocker at all. I guess my point is, generally using the variable comparison is encouraged.

Reply:
Funny enough, in JS I would use `===`, too.

Reply:
I'm fine with this change. Personally, I might lean towards

```suggestion
	if ( $before_block_visitor || $after_block_visitor ) {
```

but I don't mind the `!== null` checks either 🤷‍♂️

Reply:
it might be worth pointing out caution when relying on these kinds of micro-benchmarks. this is not only because they skew the performance picture, since we're taking the code outside of the context in which it's run, but it's easy to accidentally misrepresent what we want to test.

what we actually tested in the link above is whether it's faster to do one thing by itself, or do that same thing and then do another thing. in this case, `null === $v` and `is_null( $v )` both compile into the same literal operation, a PHP opcode called `TYPE_CHECK` as you can see in the VLD output of the code. the difference being in the two constructions is that `null !== $v` performs the type and then calls JMPZ directly, whereas `! is_null( $v )` negates the return value before calling JMPZ. why does it negate it? because we asked it to. the first branch asks PHP "tell me if this value is anything but NULL" while the second branch asks it "tell me if this value is NULL." it was able to recognize the optimization in the first case but not the second, because we've added an external data dependency on the return value through the `!`.

if you instead compare `null === $v` with `is_null( $v )` you will see that they are literally identical and there's no performance win either way. in both of these cases it asks PHP directly, "tell me if this value is NULL type."

https://3v4l.org/Daah2

in fact, look at what happens if we change `if ( ! is_null() )` into `if ( is_null() ) {} else { … }`!

https://3v4l.org/QiHIA

the two are the same speed again. but we shouldn't take away from this that…

```php
if ( is_null( $v ) ) {
    // WordPress ricer.
} else {
    do_something();
}
```

…is how we should code. I think we should take away from this that synthetic microbenchmarks are extremely unreliable indicators of code performance.

in fact, one more example of where these things can go wrong directly shows how performance can change when there is more context available. instead of calling `null !== $v` let's call `null !== $v && $v > 5`

https://3v4l.org/NJRum

now the `null !== $v` is only faster than `! is_null( $v )` some of the time, and statistically there's not sufficient evidence to demonstrate that the difference we measure isn't explained by the noisiness of our testing environment. (if you aren't seeing each branch trade places for which is faster, re-load the page and run the experiment multiples times; you should see flips).

once we add more code or real code then things can be funny in the CPU, things can happen out of order, conditions inside expressions can be pre-computed and available by the time we "get" to the `if` statement, and the more actual stuff that's happening, the harder it is to predict what matters in situations like this. micro-synthetic benchmarks are great at showing situations we will never see in production and often only show their results the way they do because we've constructed a situation specifically to lead to that outcome, unintentionally.

---

all I meant through this digression is that I support coding first to what feels most expressive and then only change that if there's is a compelling argument from intuition, experience, or measurements that the code in question behaves in a way that merits optimizations.

there's a note in the WordPress coding standards and in the PHP docs (in a comment about pre-increment being faster than post-increment and that's just demonstrably not true and it's even more trivial than here with `! is_null()` to find those counter cases.

this specific case is fairly non-controversial, because they both seem fine to me, and I even prefer `null !==` for my own style, but this is what it is - a style argument, not a performance argument. it _could_ be different if the distinction is between calling a function or an opcode, but here the generated code in both situations is nearly identical except for a negation.

thanks everyone for being non-hyped about this. I know you've all shared even-keeled thoughts and nobody is expressing strong claims or controversy; that's the same tone I'm trying to share here. only sharing this long comment because I've seen this run to micro-benchmarks multiple times and I hope that we can all understand that they rarely reflect the performance implications of the code we discuss, that they give the wrong impression just as often as they give the right one, unless the performance impact is so great that there's no wiggle room.

Reply:
100% agreed. I think such feedback is often purely code style preference rather than performance related (at least it is the case here), which is why I clarified this to not be a blocker at all. IMO not calling a function for something as simple as a strict variable comparison makes sense to me as a coding principle, but that is not driven by performance.

There are a few instances where such nit-picks have more of a performance background (e.g. using `isset()` over `array_key_exists()`), but even there the impact is negligible. On that example I would feel slightly stronger about it (since `isset()` is a language construct and thus encouraged over `array_key_exists()`) - but realistically even that is certainly not going to make or break performance in a PR, it's more of a best practice.

All that is to say: I fully agree with you, and most such comments should be addressed based on preferences between the PR author and reviewer, but not be the foundation for blocking a PR from commit.

Reply:
thanks @felixarntz - I do want to affirm what you're saying though, and note here that `is_null()` is also not an actual function but another language construct, and that's why these two styles are approximately equal. it makes me sad that in certain cases function calls still have high overhead.

sometimes the minutia does matter though. replacing arrays with a "record" style class in the Tag Processor led to an incredible performance improvement. thankfully in these cases a real impact appears in realistic testing environments.

**edit** the funnier thing to me is not that we're comparing a function call against equality, but in this case it's the `!==` which isn't what it appears to be, as that equality is hiding a call to an internal PHP function to return the type of the variable, matching the behavior of `is_null()` because it's faster than testing for equality 🙃